AI #1582: Handle for HTML/PDF anchor links for columns used in multiple datasets

mike-finopsorg opened the pull request:
#1582

Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Handle for HTML/PDF anchor links for columns used in multiple datasets #1582
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Correct links between deprecated and new field
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg committed:
Merge branch 'working_draft' into 1582-ai-handle-for-htmlpdf-anchor-links-for-columns-used-in-multiple-datasets

ijurica commented:
We should probably address the links in Metadata sections in this PR as well, right?

mike-finopsorg committed:
Fix up links in the recency metadata section
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg commented:
> We should probably address the links in Metadata sections in this PR as well, right?

I have fixed up the links I have found not working in this section. if you find more please let me know where they are so I can address them directly.

ijurica commented:
> > We should probably address the links in Metadata sections in this PR as well, right?
> 
> I have fixed up the links I have found not working in this section. if you find more please let me know where they are so I can address them directly.

@mike-finopsorg Could you please check these lines/links? It seems that they all require the appropriate prefix.

I've used this regex: `\(#(?!glossary|metadata)` in the `../FOCUS_Spec/specification/metadata` subfolder

``` 
metadata_overview.md:
   
   9: | [Data Generator](#datagenerator)     | Describes the entity delivering the dataset artifact.               |
  10: | [Dataset Instance](#datasetinstance) | Describes the nature of the dataset artifact.                       |
  11: | [Recency](#recency)                  | Describes the recency and completeness of data within the artifact. |
  12: | [Schema](#schema)                    | Describes the schema of data within the artifact.                   |  

data_generator\data_generator_overview.md:

  22  
  23: For an example of the FOCUS Data Generator metadata please refer to: [Data Generator Example](#datageneratormetadata).
  24  

data_generator\data_generator.md:

   7: * DataGenerator MUST be present in the [Data Generator](#datagenerator) object.

   9: * DataGenerator MUST conform to [StringHandling](#stringhandling) requirements.

dataset_instance\dataset_instance_id.md:

  7: * DatasetInstanceId MUST be present in an object within the [DatasetInstance](#datasetinstance) collection.

dataset_instance\dataset_instance_name.md:

  7: * DatasetInstanceName MUST be present in an object within the [DatasetInstance](#datasetinstance) collection.

dataset_instance\dataset_instance_overview.md:

  26: For an example of the FOCUS dataset instance metadata, please refer to: [Dataset Instance Metadata Example](#datasetinstancemetadata).

dataset_instance\focus_dataset.md:
 
   7: * FocusDatasetId MUST be present in an object within the [DatasetInstance](#datasetinstance) collection.

recency\dataset_instance_complete.md:

  7: * DatasetInstanceComplete MUST be present in an object within the [Recency](#recency) collection when the dataset instance is not a time-series dataset.

recency\dataset_instance_id.md:
  6  
  7: * DatasetInstanceId MUST be present in an object within the [Recency](#recency) collection.
  8  * DatasetInstanceId MUST be of type String.

recency\dataset_instance_last_updated.md:
   6  
   7: * DatasetInstanceLastUpdated MUST be present in an object within the [Recency](#recency) collection.

   9: * DatasetInstanceLastUpdated MUST conform to [DateTimeFormat](#date/timeformat) requirements.

  27: | Value format  | [Date/Time Format](#date/timeformat) |


recency\recency_last_updated.md:

  10: * RecencyLastUpdated MUST conform to [DateTimeFormat](#datetimeformat).

  27: | Value format  | [Date/Time Format](#date/timeformat)      |

recency\recency_overview.md:

  13: * Recency collection MAY contain one and only one object for every [DatasetInstance](#datasetinstance) object.
  14  * Recency collection object MUST be associated with one and only one DatasetInstance object.

recency\time_sectors\time_sector_complete.md:

  7: * TimeSectorComplete MUST be present in the [TimeSectors](#timesectors) subsection of the [Recency](#recency) metadata section.

recency\time_sectors\time_sector_end.md:

   7: * TimeSectorEnd MUST be present in the [TimeSectors](#timesectors) subsection of the [Recency](#recency) metadata section.

  10: * TimeSectorEnd MUST conform to [DateTimeFormat](#date/timeformat).

  29: | Value format  | [Date/Time Format](#date/timeformat) |

recency\time_sectors\time_sector_last_updated.md:

   7: * TimeSectorLastUpdated MUST be present in [TimeSectors](#timesectors) subsection of the [Recency](#recency) metadata section.

  10: * TimeSectorLastUpdated MUST conform to [DateTimeFormat](#date/timeformat).

  27: | Value format  | [Date/Time Format](#date/timeformat) |

recency\time_sectors\time_sector_start.md:
   6  
   7: * TimeSectorStart MUST be present in the [TimeSectors](#timesectors) subsection of the [Recency](#recency) metadata section.

  10: * TimeSectorStart MUST conform to [DateTimeFormat](#date/timeformat).

  27: | Value format  | [Date/Time Format](#date/timeformat) |

recency\time_sectors\time_sectors_overview.md:

  9: * TimeSectors MUST be present in an object within the [Recency](#recency) collection when the associated *FOCUS dataset* is defined as a time series dataset.

schema\creation_date.md:

   7: * CreationDate MUST be present in an object within the [Schema](#schema) collection.

   9: * CreationDate MUST conform to [DateTimeFormat](#date/timeformat).

  27: | Value format  | [Date/Time Format](#date/timeformat) |

schema\data_generator_version.md:

   7: * DataGeneratorVersion MAY be present in an object within the [Schema](#schema) collection.

   9: * DataGeneratorVersion MUST conform to [StringHandling](#stringhandling) requirements.

  11: * DataGeneratorVersion MUST be changed when [FocusVersion](#focusversion) is changed.

schema\dataset_instance_id.md:

  7: * DatasetInstanceId MUST be present in an object within the [Schema](#schema) collection.

schema\focus_version.md:

   7: * FocusVersion MUST be present in an object within the [Schema](#schema) collection.

schema\schema_id.md:
 
  7: * SchemaId MUST be present in an object within the [Schema](#schema) collection.

schema\schema_overview.md:

  12: * Schema collection MUST contain at least one object for every [DatasetInstance](#datasetinstance) object.
  
  26: * [Adding a new column](#addingnewcolumns)
  27: * [Removing a column](#removingcolumns)
  28: * [Changing column metadata](#changingcolumnmetadata)
  29: * [FOCUS Version has changed](#focusversionchanged)
  30: * [Data Generator Version has changed](#schemametadatatofocusdatareference)
  31: * [Correcting schema metadata errors](#providermetadataerrorcorrection)
  32  
  33: For an example of the FOCUS schema metadata, please refer to: [Schema Metadata Example](#schemametadata).

schema\column_definition\column_definition_overview.md:

  9: * ColumnDefinition MUST be present in an object within the [Schema](#schema) collection.

schema\column_definition\columnname.md:

  7: * ColumnName MUST be present in an object within the [ColumnDefinition](#columndefinition) collection.

schema\column_definition\datatype.md:

  7: * DataType MUST be present in an object within the [ColumnDefinition](#columndefinition) collection.

schema\column_definition\deprecated.md:
  
  3: The deprecation status of a column in a [Dataset Instance](#datasetinstance).

  7: * Deprecated MUST be present in an object within the [ColumnDefinition](#columndefinition) collection when the column is planned for removal.

schema\column_definition\numberscale.md:

   7: * NumberScale SHOULD be present in an object within the [ColumnDefinition](#columndefinition) collection when the column is of Decimal data type.

   9: * NumberScale MUST conform to [NumericFormat](#numericformat) requirements.

  27: | Value format  | [Numeric Format](#numericformat) |

schema\column_definition\numericprecision.md:

   7: * NumberPrecision SHOULD be present in an object within the [ColumnDefinition](#columndefinition) collection when the column is of Decimal data type.

  26: | Value format  | [Numeric Format](#numericformat) |

schema\column_definition\previouscolumnname.md:
 
  7: * PreviousColumnName MUST be present in an object within the [ColumnDefinition](#columndefinition) collection when the column was renamed.

schema\column_definition\providertagprefixes.md:

  3: The Provider Tag Prefixes define the list of prefixes used in the tag name of provider-defined [tags](#tags). This metadata is useful for the consumer to identify which tags are provider-defined vs user-defined.

  7: * ProviderTagPrefixes MUST be present in an object within the [ColumnDefinition](#columndefinition) collection when [ColumnName](#columnname) is "Tags". 

schema\column_definition\stringencoding.md:

  7: * StringEncoding MUST be present in an object within the [ColumnDefinition](#columndefinition) collection when this information is required in order to successfully read the data.

schema\column_definition\stringmaxlength.md:

   7: * StringMaxLength SHOULD be present in an object within the [ColumnDefinition](#columndefinition) collection when the column is of String data type.

  26: | Value format  | [Numeric Format](#numericformat) |

```

mike-finopsorg committed:
Pile more link corrections, mostly in manually inputted links inside metadata and appendix
Signed-off-by: Mike Fuller <mike@finops.org>

mike-finopsorg commented:
> @mike-finopsorg Could you please check these lines/links? It seems that they all require the appropriate prefix.

OK I have done another pass and more testing. I think I am getting to the end of the links that need correcting now. 🤞

shawnalpay committed:
Merge branch 'working_draft' into 1582-ai-handle-for-htmlpdf-anchor-links-for-columns-used-in-multiple-datasets

Matt-Cowsert commented on specification/datasets/cost_and_usage/columns/serviceprovidername.md:
```diff
@@ -41,4 +42,4 @@ The name of the entity that made the *resources* or *services* available for pur
 ## Introduced (version)
 
-1.3 Introduced as a replacement for [ProviderName](#providername)
+1.3 Introduced as a replacement for [ProviderName](#datasets.costandusage.provider-deprecated)
```
```suggestion
1.3 Introduced as a replacement for [ProviderName](#datasets.costandusage.providername)
```

Matt-Cowsert replied:
while the column is marked for deprecation, the file name, as far as I can tell, hasn't changed.

Matt-Cowsert commented on specification/datasets/cost_and_usage/columns/providername.md:
```diff
@@ -41,4 +41,4 @@ The name of the entity that made the *resources* or *services* available for pur
 ## Deprecated (version)
 
-1.3 Replaced by [ServiceProviderName](#serviceprovidername)
+1.3 Replaced by [ServiceProviderName](#datasets.costandusage.provider-deprecated)
```
```suggestion
1.3 Replaced by [ServiceProviderName](#datasets.costandusage.serviceprovidername)
```

Matt-Cowsert replied:
I expect we need to point to the service provider name file vs. provider-deprecated (not a file I can find).

Matt-Cowsert reviewed (changes requested):
Two quick adjustments (queued as suggestions for ease) on the ServiceProvider and Provider files.

shawnalpay committed:
Apply suggestions from code review

Co-authored-by: Matt Cowsert <matthew@finops.org>

shawnalpay commented:
Good catches, @Matt-Cowsert!  I've committed both of those and checked the PDF to ensure they work as expected.

jpradocueva commented:
Approved by the group without objections or abstentions during the January 8th Members’ call.

mike-finopsorg committed:
AI #1583: Added Markdown-PP option to skip headers in the TOC (#1671)

PR to merge into
https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1582

Signed-off-by: Mike Fuller <mike@finops.org>
